Source/UZKArchive.m

```diff
@@ -419,66 +419,72 @@ - (BOOL)extractFilesTo:(NSString *)destinationDirectory
-                                            error:&strongError];
-                    
-                    if (!data && strongError) {
-                        [self assignError:&strongError code:UZKErrorCodeFileRead
-                                   detail:[NSString localizedStringWithFormat:NSLocalizedString(@"Error reading current file '%@'", @"Detailed error string"),
-                                           info.filename]];
-                        return;
-                    }
-                    
-                    int err = unzCloseCurrentFile(self.unzFile);
-                    if (err != UNZ_OK) {
-                        if (err == UZKErrorCodeCRCError) {
-                            err = UZKErrorCodeInvalidPassword;
-                        }
+                    					
+					BOOL isDirectory = YES;
+					NSString *extractDir = extractPath.stringByDeletingLastPathComponent;
+					if (![fm fileExistsAtPath:extractDir]) {
+						BOOL directoriesCreated = [fm createDirectoryAtPath:extractDir
+												withIntermediateDirectories:YES
+																 attributes:nil
+																	  error:error];
+						if (!directoriesCreated) {
+							[self assignError:&strongError code:UZKErrorCodeOutputError
+									   detail:[NSString localizedStringWithFormat:NSLocalizedString(@"Failed to create destination directory: %@", @"Detailed error string"),
+											   extractDir]];
+							return;
+						}
+					} else if (!isDirectory) {
+						[self assignError:&strongError code:UZKErrorCodeOutputErrorPathIsAFile
+								   detail:[NSString localizedStringWithFormat:NSLocalizedString(@"Extract path exists, but is not a directory: %@", @"Detailed error string"),
+										   extractDir]];
+						return;
+					}
+
+					
+					NSURL *defalatedDirectoryURL = [NSURL fileURLWithPath:destinationDirectory];
+					NSURL *deflatedFileURL = [defalatedDirectoryURL URLByAppendingPathComponent:info.filename];
+					NSString *path = deflatedFileURL.path;
+					
+					BOOL createSuccess = [fm createFileAtPath:path
+													 contents:nil
+												   attributes:nil];
+
+					if (!createSuccess) {
+						[self assignError:&strongError code:UZKErrorCodeOutputError
+								   detail:[NSString localizedStringWithFormat:NSLocalizedString(@"Error creating current file (%d) '%@'", @"Detailed error string"),
+										   strongError, info.filename]];
+						return;
+					}
+					
+					__block double dataLength = 0;
+					
+					NSError *handleError = nil;
+					NSFileHandle *deflated = [NSFileHandle fileHandleForWritingToURL:deflatedFileURL
+																			   error:&handleError];
+
+					
+					BOOL extractSuccess = [self extractBufferedDataFromFile:info.filename
+																  error:&strongError
+																 action:
+									^(NSData *dataChunk, CGFloat percentDecompressed) {
+										dataLength += dataChunk.length;
+										[deflated writeData:dataChunk];
```

Comment:
It would be great to report updated progress here, since we now have more granular information. Instead of incrementing `bytesDecompressed` at the end of the loop, it would be cool to see the following:

```objc
bytesDecompressed += dataChunk.length;
if (progress) {
    progress(info, bytesDecompressed / totalSize.doubleValue);
}
```